sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -197,4 +204,26 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
 
     case _ => Nil
   }
+
+  private def getBlockLocations(file: FileStatus): Array[BlockLocation] = file match {
+    case f: LocatedFileStatus => f.getBlockLocations
+    case f => Array.empty[BlockLocation]
+  }
+
+  // Given locations of all blocks of a single file, `blockLocations`, and an `offset` position of
+  // the same file, find out the index of the block that contains this `offset`. If no such block
+  // can be found, returns -1.
+  private def getBlockIndex(blockLocations: Array[BlockLocation], offset: Long): Int = {
+    blockLocations.indexWhere { b =>
+      b.getOffset <= offset && offset < b.getOffset + b.getLength
+    }
+  }
+
+  // Given locations of all blocks of a single file, `blockLocations`, and an `offset` position of
+  // the same file, find out the block that contains this `offset`, and returns location hosts of
+  // that block. If no such block can be found, returns an empty array.
+  private def getBlockHosts(blockLocations: Array[BlockLocation], offset: Long): Array[String] = {
```

Comment:
Should we pick the hosts based on range that the host hold most of data in that range?